content/client-lib-development-guide/features.textile

```diff
@@ -468,7 +468,7 @@ h3(#realtime-presence). Presence
 * @(RTP1)@ When a channel @ATTACHED@ @ProtocolMessage@ is received, the @ProtocolMessage@ may contain a @HAS_PRESENCE@ bit flag indicating that there are currently members present on the channel, see "TR3":#TR3 . If the flag is 1, the server will shortly perform a @SYNC@ operation as described in "RTP18":#RTP18 . If that flag is 0 or there is no @flags@ field, the presence map should be considered in sync immediately with no members present on the channel
 * @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
 ** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
-** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of an incoming presence message corresponds to its @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
+** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of both presence message corresponds to the @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
```

Comment:
> Well we could simply use `msgSerial` - ie include that in a `PresenceMessage` - if we also knew that no two `PresenceMessage`s with the same `clientId` would be sent in the same `ProtocolMessage`.

Does realtime guarantee to process protocol messages atomically, or could it be made to do so? I.e. if I send a protocol message with two presence messages 'foo enter' and 'foo leave', can it guarantee that no-one will get a SYNC showing a presence set with foo in it?

If so, we could get away with only using `msgSerial` (plus timestamp for synthesized msgs), since SYNCs will only have data from the highest-index presence message for that clientId in the protocol message, and while the PRESENCE messages will have both, we can specify the newness comparison uses >= rather than > and that clients process the presence array in order, so higher-index presence message would just overwrite lower.